### ISSUE TITLE
osu_iallgather at 32 ranks hits a memory registration cache problem

### ISSUE THREAD

**Issue by hppritcha** [ofi-cray/libfabric-cray#604]
When running the osu_iallgather test (linked against Open MPI with libfabric OFI enabled) its easy to hit a problem with what appears to be duplicate key problem in the cache.

`srun --label --exclusive -N 2 --np 32 ./osu_iallgather`

Here's the output and traceback from gdb of core file:

n17276@tiger:/cray/css/users/n17276/osu-micro-benchmarks-5.0/mpi/collective> (master *#)srun -N 2 -n 32 --label --exclusive ./osu_iallgather 

```
00: 
00: # OSU MPI Non-blocking Allgather Latency Test v5.0
00: # Overall = Coll. Init + Compute + MPI_Test + MPI_Wait
00: 
00: # Size           Overall(us)       Compute(us)    Pure Comm.(us)        Overlap(%)
00: 1                     403.78            206.40            200.19              1.40
00: 2                     398.34            201.45            195.99              0.00
00: 4                     398.34            201.41            195.94              0.00
00: 8                     398.29            201.26            195.87              0.00
00: 16                    398.52            201.50            196.11              0.00
00: 32                    403.06            203.75            198.37              0.00
00: 64                    403.46            204.01            198.61              0.00
00: 128                   434.33            219.68            213.77              0.00
00: 256                   436.77            220.92            215.04              0.00
00: 512                   445.52            225.58            219.44              0.00
00: 1024                  484.98            245.81            239.34              0.07
00: 2048                  621.20            313.23            305.43              0.00
00: 4096                  918.78            486.54            474.67              8.94
00: 8192                 1693.24            884.27            863.25              6.29
30: found dup key 0x00007fffaf80e000:262144, in tree 0x00007fffaf80e000:16384
30: found dup key 0x00007fffaf80e000:262144, in tree 0x00007fffaf80e000:16384
30: osu_iallgather: prov/gni/src/gnix_mr.c:1217: __mr_cache_search_stale: Assertion `rc == RBT_STATUS_OK' failed.
30: [nid00013:17765] *** Process received signal ***
30: [nid00013:17765] Signal: Aborted (6)
30: [nid00013:17765] Signal code:  (-6)
30: [nid00013:17765] [ 0] /lib64/libpthread.so.0(+0xf850)[0x7ffff6a9a850]
30: [nid00013:17765] [ 1] /lib64/libc.so.6(gsignal+0x35)[0x7ffff6741885]
30: [nid00013:17765] [ 2] 
30: /lib64/libc.so.6(abort+0x181)[0x7ffff6742e61]
30: [nid00013:17765] [ 3] /lib64/libc.so.6(__assert_fail+0xf0)[0x7ffff673a740]
30: [nid00013:17765] [ 4] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x65e5c)[0x7ffff5992e5c]
30: [nid00013:17765] [ 5] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x66cdb)[0x7ffff5993cdb]
30: [nid00013:17765] [ 6] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x61feb)[0x7ffff598efeb]
30: [nid00013:17765] [ 7] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x69f4f)[0x7ffff5996f4f]
30: [nid00013:17765] [ 8] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x8d562)[0x7ffff59ba562]
30: [nid00013:17765] [ 9] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x8db64)[0x7ffff59bab64]
30: [nid00013:17765] [10] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x8f06a)[0x7ffff59bc06a]
30: [nid00013:17765] [11] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x75232)[0x7ffff59a2232]
30: [nid00013:17765] [12] /home/users/n17276/libfabric_install/lib/libfabri
30: c.so.1(+0x4583a)[0x7ffff597283a]
30: [nid00013:17765] 
30: [13] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x473b8)[0x7ffff59743b8]
30: [nid00013:17765] [14] /home/users/n17276/libfabric_install/lib/libfabric.so.1(+0x4753c)[0x7ffff597453c]
30: [nid00013:17765] [15] 
30: /home/users/n17276/ompi_install/lib/libmpi.so.0(ompi_mtl_ofi_progress_no_inline+0x57)[0x7ffff6e259d7]
30: [nid00013:17765] [16] /home/users/n17276/ompi_install/lib/libopen-pal.so.0(opal_progress+0x2a)[0x7ffff5c24aea]
30: [nid00013:17765] [17] 
30: /home/users/n17276/ompi_install/lib/libmpi.so.0(ompi_request_default_wait+0x55)[0x7ffff6d1e5c5]
30: [nid00013:17765] [18] 
30: /home/users/n17276/ompi_install/lib/libmpi.so.0(MPI_Wait+0x5d)[0x7ffff6d556ed]

Program terminated with signal 6, Aborted.
#0  0x00007ffff6741885 in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff6741885 in raise () from /lib64/libc.so.6
#1  0x00007ffff6742df5 in abort () from /lib64/libc.so.6
#2  0x00007ffff673a740 in __assert_fail () from /lib64/libc.so.6
#3  0x00007ffff5992e5c in __mr_cache_search_stale (cache=0x7b0800, domain=0x7421a0, address=140736137912320, length=16384, dst_cq_hndl=0x0, 
    flags=0, vmdh_index=4294967295, entry=0x7fffffff6c58, key=0x7fffffff6c60) at prov/gni/src/gnix_mr.c:1217
#4  0x00007ffff5993cdb in __mr_cache_register (cache=0x7b0800, domain=0x7421a0, address=140736137912320, length=16384, dst_cq_hndl=0x0, flags=0, 
    vmdh_index=4294967295, mr=0x7fffffff6d28) at prov/gni/src/gnix_mr.c:1374
#5  0x00007ffff598efeb in gnix_mr_reg (fid=0x7421a0, buf=0x7fffaf81e000, len=16384, access=768, offset=0, requested_key=0, flags=0, 
    mr_o=0x7fffffff6ef0, context=0x0) at prov/gni/src/gnix_mr.c:683
#6  0x00007ffff5996f4f in __gnix_rndzv_req (arg=0x747f40) at prov/gni/src/gnix_msg.c:405
#7  0x00007ffff59ba562 in __gnix_vc_push_work_reqs (vc=0x7abd00) at prov/gni/src/gnix_vc.c:1631
#8  0x00007ffff59bab64 in __gnix_nic_vc_work_progress (nic=0x763cd0) at prov/gni/src/gnix_vc.c:1691
#9  0x00007ffff59bc06a in _gnix_nic_vc_progress (nic=0x763cd0) at prov/gni/src/gnix_vc.c:1914
#10 0x00007ffff59a2232 in _gnix_nic_progress (nic=0x763cd0) at prov/gni/src/gnix_nic.c:547
#11 0x00007ffff597283a in __gnix_cq_progress (cq=0x76b800) at prov/gni/src/gnix_cq.c:262
#12 0x00007ffff59743b8 in gnix_cq_readfrom (cq=0x76b800, buf=0x7fffffff72c0, count=1, src_addr=0x0) at prov/gni/src/gnix_cq.c:498
#13 0x00007ffff597453c in gnix_cq_read (cq=0x76b800, buf=0x7fffffff72c0, count=1) at prov/gni/src/gnix_cq.c:530
#14 0x00007ffff6e259d7 in ompi_mtl_ofi_progress_no_inline () from /home/users/n17276/ompi_install/lib/libmpi.so.0
#15 0x00007ffff5c24aea in opal_progress () from /home/users/n17276/ompi_install/lib/libopen-pal.so.0
#16 0x00007ffff6d1e5c5 in ompi_request_default_wait () from /home/users/n17276/ompi_install/lib/libmpi.so.0
#17 0x00007ffff6d556ed in PMPI_Wait () from /home/users/n17276/ompi_install/lib/libmpi.so.0
#18 0x0000000000401574 in main (argc=1, argv=0x7fffffff75a8) at osu_iallgather.c:110

```

**Comment by jswaro** [comment]
Looks like the mr cache contains an entry that is not supposed to be there. There is an entry that should have been subsumed, but it is still in the tree. Either the tree was not properly traversed or the logic for removal is broken.

**Comment by jswaro** [comment]
closed by #622
